central/scannerdefinitions/handler/handler.go

```diff
@@ -388,81 +400,103 @@ func (h *httpHandler) openOfflineBlob(ctx context.Context, blobName string) (*vu
 	}
-	log.Debugf("Compressed data file is available: %s", openedFile.Name())
+	log.Debugf("Online data %s file %s is available: %s", t, opts.urlPath, online.Name())
+
+	var success bool
+	f := &vulDefFile{File: online, modTime: modTime}
+	defer func() {
+		// If we fail from here, then we should close "online". Similarly, if the client desired a file inside "online",
+		// then no need to keep "online" open any longer.
+		if !success || f.File != online {
+			utils.IgnoreError(online.Close)
+		}
+	}()
+
 	switch t {
 	case v2UpdaterType:
 		if opts.fileName == "" {
-			return &vulDefFile{File: openedFile, modTime: onlineTime}, nil
+			break
 		}
+		// We want to find a specific file inside the toplevel online file.
 		fallthrough
 	case mappingUpdaterType:
-		// openFromArchive will copy the contents of opts.fileName from openedFile to
-		// targetFile. Because of this, openedFile is not needed outside this function,
-		// so close it here.
-		defer utils.IgnoreError(openedFile.Close)
-		targetFile, _, err := h.openFromArchive(openedFile.Name(), opts.fileName)
+		targetFile, _, err := h.openFromArchive(online.Name(), opts.fileName)
 		if err != nil {
 			return nil, err
 		}
-		return &vulDefFile{File: targetFile, modTime: onlineTime}, nil
+		f.File = targetFile
 	case vulnerabilityUpdaterType:
-		return &vulDefFile{File: openedFile, modTime: onlineTime}, nil
+		// Always return the toplevel online file.
+	default:
+		return nil, fmt.Errorf("unknown updater type: %s", t)
 	}
-	return nil, fmt.Errorf("unknown Scanner V4 updater type: %s", t)
+
+	success = true
+	return f, nil
 }
 
 // getUpdater gets or creates an updater for the scanner definitions identified
 // by the given updater type and a URL path to the definitions file. If the
 // updater was created, it is no started here, callers are expected to start it.
-func (h *httpHandler) getUpdater(t updaterType, urlPath string) *requestedUpdater {
+func (h *httpHandler) getUpdater(t updaterType, urlPath string) (*requestedUpdater, error) {
 	h.updatersLock.Lock()
 	defer h.updatersLock.Unlock()
 
 	fileName := strings.ReplaceAll(filepath.Join(t.String(), urlPath), "/", "-")
-	updater, exists := h.updaters[fileName]
+	u, exists := h.updaters[fileName]
 	if !exists {
 		var updateURL *url.URL
-		var ext string
 		switch t {
 		case mappingUpdaterType:
 			updateURL = scannerUpdateBaseURL.JoinPath(scannerV4MappingSubDir, scannerV4MappingFile)
-			ext = ".zip"
 		case vulnerabilityUpdaterType:
 			updateURL = scannerUpdateBaseURL.JoinPath(scannerV4VulnSubDir, urlPath)
-			ext = ".json.zst"
-		default: // uuid
+		case v2UpdaterType:
 			updateURL = scannerUpdateBaseURL.JoinPath(urlPath, scannerV2DiffFile)
-			ext = ".zip"
+		default:
+			return nil, fmt.Errorf("unknown updater type: %s", t)
 		}
 		filePath := filepath.Join(h.dataDir, fileName)
-		// Use a default extension if the URL path does not contain one.
-		if filepath.Ext(fileName) == "" {
-			filePath += ext
-		}
```

Comment:
Nit: In the container this would quickly tell use which type the file is when manually inspecting its contents/debugging. I am guessing you removed it because it didn't look like it was relevant/important?

Reply:
yeah I didn't think this added too much value. Also the `json.zst` which was previously hardcoded may no longer be true, so we'd have to change the logic to account for that potentially (probably) being a zip. This way, the code seemed cleaner, too.

Would you rather I add this back?

Reply:
Nah, the cleaner code surpasses the debugging convenience. Thanks for clarifying.